libp2p/io/msgio.py

```diff
@@ -5,80 +5,85 @@
 
 NOTE: currently missing the capability to indicate lengths by "varint" method.
 """
-# TODO unify w/ https://github.com/libp2p/py-libp2p/blob/1aed52856f56a4b791696bbcbac31b5f9c2e88c9/libp2p/utils.py#L85-L99  # noqa: E501
-from typing import Optional, cast
+from abc import abstractmethod
 
-from libp2p.io.abc import Closer, ReadCloser, Reader, ReadWriteCloser, WriteCloser
+from libp2p.io.abc import MsgReadWriteCloser, Reader, ReadWriteCloser
 from libp2p.io.utils import read_exactly
+from libp2p.utils import decode_uvarint_from_stream, encode_varint_prefixed
+
+from .exceptions import MessageTooLarge
 
-SIZE_LEN_BYTES = 4
 BYTE_ORDER = "big"
 
 
-async def read_length(reader: Reader) -> int:
-    length_bytes = await read_exactly(reader, SIZE_LEN_BYTES)
+async def read_length(reader: Reader, size_len_bytes: int) -> int:
+    length_bytes = await read_exactly(reader, size_len_bytes)
     return int.from_bytes(length_bytes, byteorder=BYTE_ORDER)
 
 
-def encode_msg_with_length(msg_bytes: bytes) -> bytes:
-    len_prefix = len(msg_bytes).to_bytes(SIZE_LEN_BYTES, "big")
+def encode_msg_with_length(msg_bytes: bytes, size_len_bytes: int) -> bytes:
```

Comment:
Why is the size passed in here instead of measuring it from `len(msg_bytes)`?

Reply:
`size_len_bytes` means the size of the length field in the message. `len(msg_bytes)` is encoded and put in the length field.